crates/api_crud/src/user/create.rs

```diff
@@ -207,34 +187,404 @@ pub async fn register(
+    oauth_user_id = serde_json::from_value::<String>(oauth_user_id_value.clone())
+      .ok()
+      .ok_or(LemmyErrorType::OauthLoginFailed)?;
+  } else {
+    return Err(LemmyErrorType::OauthLoginFailed)?;
+  }
+
+  let mut login_response = LoginResponse {
+    jwt: None,
+    registration_created: false,
+    verify_email_sent: false,
+  };
+
+  // Lookup user by oauth_user_id
+  let mut local_user_view =
+    LocalUserView::find_by_oauth_id(&mut context.pool(), oauth_provider.id, &oauth_user_id).await?;
+
+  let local_user: LocalUser;
+  if let Some(user_view) = local_user_view {
+    // user found by oauth_user_id => Login user
+    local_user = user_view.clone().local_user;
+
+    check_user_valid(&user_view.person)?;
+    check_email_verified(&user_view, &site_view)?;
+    check_registration_application(&user_view, &site_view.local_site, &mut context.pool()).await?;
+  } else {
+    // user has never previously registered using oauth
+
+    // prevent registration if registration is closed
+    if local_site.registration_mode == RegistrationMode::Closed {
+      Err(LemmyErrorType::RegistrationClosed)?
+    }
+
+    // prevent registration if registration is closed for OAUTH providers
+    if !local_site.oauth_registration {
+      return Err(LemmyErrorType::OauthRegistrationClosed)?;
+    }
+
+    // Extract the OAUTH email claim from the returned user_info
+    let email: String;
+    if let Some(email_value) = user_info.get("email") {
+      email = serde_json::from_value::<String>(email_value.clone())
+        .ok()
+        .ok_or(LemmyErrorType::OauthLoginFailed)?;
+    } else {
+      return Err(LemmyErrorType::OauthLoginFailed)?;
+    }
+
+    let require_registration_application =
+      local_site.registration_mode == RegistrationMode::RequireApplication;
+
+    // Lookup user by OAUTH email and link accounts
+    local_user_view = LocalUserView::find_by_email(&mut context.pool(), &email).await?;
+
+    let person;
+    if let Some(user_view) = local_user_view {
+      // user found by email => link and login if linking is allowed
+
+      // we only allow linking by email when email_verification is required otherwise emails cannot
+      // be trusted
+      if oauth_provider.account_linking_enabled && site_view.local_site.require_email_verification {
+        // WARNING:
+        // If an admin switches the require_email_verification config from false to true,
+        // users who signed up before the switch could have accounts with unverified emails falsely
+        // marked as verified.
+
+        // make sure the user is valid before linking
+        check_user_valid(&user_view.person)?;
+
+        // make sure the email is verified before linking
+        check_email_verified(&user_view, &site_view)?;
+
+        // make sure the registration application is approved or not required before linking
+        check_registration_application(&user_view, &site_view.local_site, &mut context.pool())
```

Comment:
No need to put comments onto each of those checks. If needed you can improve docs on each function definition.